### PR TITLE
Csv Filter with test Coverage for GoodsReceipt

### PR DESCRIPTION
**Pull request by stephaneleong**
#87 and #79

### CODE REVIEW

File: src/main/java/seedu/address/storage/CsvGoodsStorage.java [hunk 1]
```diff
@@ -43,7 +52,44 @@ public Optional<ReadOnlyReceiptLog> readGoods(Path filePath) throws DataLoadingE
...
+                    Date arrivalDate = new Date(trimmedLine[0].replace("T", " "));
+                    GoodsName name = new GoodsName(goodDetails[0]);
+                    GoodsCategories goodsCategory = GoodsCategories.valueOf(goodDetails[1]);
+                    boolean isDelivered = Boolean.parseBoolean(trimmedLine[2]);
+                    double price = Double.parseDouble(trimmedLine[3]);
+                    Date procurementDate = new Date(trimmedLine[4].replace("T", " "));
+                    int quantity = Integer.parseInt(trimmedLine[5]);
+                    Name supplier = new Name(trimmedLine[6]);
+
+                    return true;
```
Comment (thenosewizard) [inline src/main/java/seedu/address/storage/CsvGoodsStorage.java:84]: Good in trying to improve error handling

File: src/main/java/seedu/address/storage/CsvGoodsStorage.java [hunk 2]
```diff
@@ -43,7 +52,44 @@ public Optional<ReadOnlyReceiptLog> readGoods(Path filePath) throws DataLoadingE
...
-        receiptLog.setReceipts(CsvUtil.readCsvFile(filePath, GoodsReceipt.class).get());
+
+        CsvToBeanFilter filter = new CsvToBeanFilter() {
+            private final HashSet<String> seenReceipts = new HashSet<>();
+
+            @Override
+            public boolean allowLine(String[] line) {
+                String[] trimmedLine = Arrays.stream(line).map(String::trim).toArray(String[]::new);
+                boolean isDuplicate = !seenReceipts.add(Arrays.toString(trimmedLine));
+                boolean hasMissingFields = trimmedLine.length != 7;
```
Comment (ossaris56) [inline src/main/java/seedu/address/storage/CsvGoodsStorage.java:63]: Perhaps the constant 7 could be assigned to a constant variable instead. Would the code easier to read.

File: src/main/java/seedu/address/storage/CsvGoodsStorage.java [hunk 3]
```diff
@@ -43,7 +52,44 @@ public Optional<ReadOnlyReceiptLog> readGoods(Path filePath) throws DataLoadingE
...
+                if (hasMissingFields || isDuplicate) {
+                    return false;
+                }
+
+                String[] goodDetails = trimmedLine[1].split(",", 2);
```
Comment (ossaris56) [inline src/main/java/seedu/address/storage/CsvGoodsStorage.java:68]: Here as well

**Review by ossaris56 — COMMENTED** [review]
LGTM! Good testing!

File: src/main/java/seedu/address/commons/util/CsvUtil.java [hunk 4]
```diff
@@ -48,6 +64,7 @@ public static <T> Optional<List<T>> readCsvFile(Path filePath, Class<T> clazz) t
         try (FileReader reader = new FileReader(filePath.toFile())) {
             List<T> beans = new CsvToBeanBuilder<T>(reader)
                     .withType(clazz)
+                    .withFilter(filter)
                     .withFieldAsNull(CSVReaderNullFieldIndicator.BOTH)
                     .build()
                     .parse();
```
Comment (yanhwee) [inline src/main/java/seedu/address/commons/util/CsvUtil.java:70]: good use of builder design pattern

File: src/test/java/seedu/address/model/goods/GoodsTest.java [hunk 5]
```diff
@@ -53,4 +53,12 @@ public void equals_differentTypes_failure() {
         GoodsName otherGoods = new GoodsName("Other Bread");
         assertNotEquals(testGoods, otherGoods);
     }
+
+    @Test
+    public void convertToCsvWrite_success() {
```
Comment (yanhwee) [inline src/test/java/seedu/address/model/goods/GoodsTest.java:58]:
```suggestion
    public void convertToCsvWrite_validArgs_success() {
```

**Review by yanhwee — COMMENTED** [review]
Generally okay, to fix some test cases names.

**Pull request merged**